repositories/src/main/scala/cromwell/pipeline/datastorage/dto/ProjectSearch.scala

```diff
@@ -0,0 +1,68 @@
+package cromwell.pipeline.datastorage.dto
+
+import play.api.libs.json._
+
+sealed trait ProjectSearchFilter
+object ProjectSearchFilter {
+  implicit lazy val projectSearchFilterReads: Reads[ProjectSearchFilter] = Reads[ProjectSearchFilter] { json =>
+    (json \ "type").validate[String].flatMap {
+      case "all"            => json.validate[AllFilter.type]
+      case "name"           => json.validate[ByNameFilter]
+      case "files"          => json.validate[ByFilesFilter]
+      case "configurations" => json.validate[ByConfigFilter]
+      case "OR"             => json.validate[ComplexOrFilter]
+      case "AND"            => json.validate[ComplexAndFilter]
+    }
+  }
+}
+
+case object AllFilter extends ProjectSearchFilter {
+  implicit lazy val allReads: Reads[AllFilter.type] = Json.reads[AllFilter.type]
+}
+
+final case class ByNameFilter(mode: NameSearchMode, value: String) extends ProjectSearchFilter
+object ByNameFilter {
+  implicit lazy val byNameReads: Reads[ByNameFilter] = Json.reads[ByNameFilter]
+}
+
+sealed trait NameSearchMode
+case object FullMatch extends NameSearchMode
+case object RegexpMatch extends NameSearchMode
+
+object NameSearchMode {
+  implicit lazy val nameSearchModeReads: Reads[NameSearchMode] = {
+    implicitly[Reads[String]].map(NameSearchMode.fromString)
+  }
+
+  def fromString(s: String): NameSearchMode = s match {
+    case "full_match"   => FullMatch
+    case "regexp_match" => RegexpMatch
+  }
+}
+
+final case class ByConfigFilter(mode: String, value: Boolean) extends ProjectSearchFilter
```

Comment:
I don't see any usages of `mode` field anywhere (same for `ByFilesFilter`). If it's not required, why do we need it? If it's required, why it's unused?

Reply:
As for now I believe we can replace {"mode" : "exists", "value" : boolean} with {"exists" : boolean} 
Maybe the idea was to add more modes for config and file search?
P.S. adt for lonely Exist mode created

Reply:
> Maybe the idea was to add more modes for config and file search?
The idea was to make addition of more modes possible in the future. So even though now it is the only mode, we might want to add more in any moment

Reply:
> Maybe the idea was to add more modes for config and file search?

The idea was to make addition of more modes possible in the future. So even though now it is the only mode, we might want to add more in any moment. What happens if tomorrow we decide to add something like this?
```json
{
  "type": "files",
  "mode: "contain",
  "value": "some_file_name"
}
```

Will addition of new adt case `Contain` cover it? Does your current case class structure allow such changes now?

Reply:
That is a major flaw in current implementation. For now addition of new mode will work fine only with specific type of `value`. It should be rethought and rebuilt